src/util/netif.cpp

```diff
@@ -0,0 +1,275 @@
+    for (nlmsghdr* hdr = (nlmsghdr*)response; NLMSG_OK(hdr, response_len); hdr = NLMSG_NEXT(hdr, response_len)) {
+        rtmsg* r = (rtmsg*)NLMSG_DATA(hdr);
+        int remaining_len = RTM_PAYLOAD(hdr);
+
+        // Iterate over the attributes.
+        rtattr *rta_gateway = nullptr;
+        int scope_id = 0;
+        for (rtattr* attr = RTM_RTA(r); RTA_OK(attr, remaining_len); attr = RTA_NEXT(attr, remaining_len)) {
+            if (attr->rta_type == RTA_GATEWAY) {
+                rta_gateway = attr;
+            } else if (attr->rta_type == RTA_OIF) {
+                Assume(sizeof(int) == RTA_PAYLOAD(attr));
+                std::memcpy(&scope_id, RTA_DATA(attr), sizeof(scope_id));
+            }
+        }
+
+        // Found gateway?
+        if (rta_gateway != nullptr) {
+            if (network == NET_IPV4) {
+                Assume(sizeof(in_addr) == RTA_PAYLOAD(rta_gateway));
+                in_addr gw;
+                std::memcpy(&gw, RTA_DATA(rta_gateway), sizeof(gw));
+                return CNetAddr(gw);
+            } else if (network == NET_IPV6) {
+                Assume(sizeof(in6_addr) == RTA_PAYLOAD(rta_gateway));
+                in6_addr gw;
+                std::memcpy(&gw, RTA_DATA(rta_gateway), sizeof(gw));
+                return CNetAddr(gw, scope_id);
+            }
+        }
+    }
+
+    return std::nullopt;
+}
+
+#elif defined(WIN32)
+
+#include <iphlpapi.h>
+
+std::optional<CNetAddr> QueryDefaultGateway(Network network)
+{
+    NET_LUID interface_luid = {};
+    SOCKADDR_INET destination_address = {};
+    MIB_IPFORWARD_ROW2 best_route = {};
+    SOCKADDR_INET best_source_address = {};
+    DWORD best_if_idx = 0;
+    DWORD status = 0;
+
+    // Pass empty destination address of the requested type (:: or 0.0.0.0) to get interface of default route.
+    Assume(network == NET_IPV4 || network == NET_IPV6);
+    if (network == NET_IPV4) {
+        destination_address.si_family = AF_INET;
+    } else if(network == NET_IPV6) {
+        destination_address.si_family = AF_INET6;
+    } else {
+        return std::nullopt;
+    }
+
+    status = GetBestInterfaceEx((sockaddr*)&destination_address, &best_if_idx);
+    if (status != NO_ERROR) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get best interface for default route: %s\n", SysErrorString(status));
+        return std::nullopt;
+    }
+
+    // Get best route to default gateway.
+    // Leave interface_luid at all-zeros to use interface index instead.
+    status = GetBestRoute2(&interface_luid, best_if_idx, nullptr, &destination_address, 0, &best_route, &best_source_address);
+    if (status != NO_ERROR) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Error, "Could not get best route for default route for interface index %d: %s\n",
+                best_if_idx, SysErrorString(status));
+        return std::nullopt;
+    }
+
+    if (network == NET_IPV4) {
+        Assume(best_route.NextHop.si_family == AF_INET);
```

Comment:
OK, you're right, this could happen: https://learn.microsoft.com/en-us/windows/win32/api/netioapi/ns-netioapi-mib_ipforward_row2

> NextHop
>
> Type: [SOCKADDR_INET](https://learn.microsoft.com/en-us/windows/desktop/api/ws2ipdef/ns-ws2ipdef-sockaddr_inet)
>
> For a remote route, the IP address of the next system or gateway en route. If the route is to a local loopback address or an IP address on the local link, the next hop is unspecified (all zeros). For a local loopback route, this member should be an IPv4 address of 0.0.0.0 for an IPv4 route entry or an IPv6 address of 0::0 for an IPv6 route entry.